core/distributed/helpers.hpp

```diff
@@ -124,5 +128,44 @@ bool is_distributed(Arg* linop, Rest*... rest)
 }
 
 
+/**
+ * Cast an input linop to the correct underlying vector type (dense/distributed)
+ * and passes it to the given function.
+ *
+ * @tparam ValueType  The value type of the underlying dense or distributed
+ * vector.
+ * @tparam T  The linop type, either LinOp, or const LinOp.
+ * @tparam F  The function type.
+ * @tparam Args  The types for the additional arguments of f.
+ *
+ * @param linop  The linop to be casted into either a dense or distributed
+ *               vector.
+ * @param f  The function that is to be called with the correctly casted linop.
+ * @param args  The additional arguments of f.
+ */
+template <typename ValueType, typename T, typename F, typename... Args>
+void vector_dispatch(T* linop, F&& f, Args... args)
+{
+#if GINKGO_BUILD_MPI
+    if (is_distributed(linop)) {
+        using type = std::conditional_t<
+            std::is_const<T>::value,
+            const experimental::distributed::Vector<ValueType>,
+            experimental::distributed::Vector<ValueType>>;
+        f(dynamic_cast<type*>(linop), std::forward<Args>(args)...);
```

Comment:
it should check whether dynamic_cast<type*>(linop) is a nullptr

Reply:
Could you expand on that? I don't think any of the two branches makes sense in that case, so I could only throw. Is that your suggestion?

Reply:
yes, it is. when the linop can not be the type, it will pass the nullptr and lead segmentation fault, I think